File: objectModel/Python/cdm/objectmodel/cdm_import.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import Optional, TYPE_CHECKING
import warnings

from cdm.utilities import logger, ResolveOptions
from cdm.enums import CdmObjectType
from cdm.enums import CdmLogCode
from cdm.utilities.string_utils import StringUtils

from .cdm_object_simple import CdmObjectSimple

if TYPE_CHECKING:
    from cdm.objectmodel import CdmCorpusContext, CdmDocumentDefinition
    from cdm.utilities import FriendlyFormatNode, VisitCallback


class CdmImport(CdmObjectSimple):
    def __init__(self, ctx: 'CdmCorpusContext', corpus_path: str, moniker: str) -> None:
        super().__init__(ctx)

        self._TAG = CdmImport.__name__
        self.corpus_path = corpus_path  # type: str
        self.moniker = moniker  # type: str

        # --- internal ---
        self._document = None  # type: Optional[CdmDocumentDefinition]

    @property
    def doc(self) -> Optional['CdmDocumentDefinition']:
        warnings.warn('This property is deprecated and it is likely to be removed soon..', DeprecationWarning)
        return self._document

    @property
    def object_type(self) -> 'CdmObjectType':
        return CdmObjectType.IMPORT

    def copy(self, res_opt: Optional['ResolveOptions'] = None, host: Optional['CdmImport'] = None) -> 'CdmImport':
        if not res_opt:
            res_opt = ResolveOptions(wrt_doc=self, directives=self.ctx.corpus.default_resolution_directives)

        if not host:
            copy = CdmImport(self.ctx, self.corpus_path, self.moniker)
        else:
            copy = host
            copy.ctx = self.ctx
            copy.corpus_path = self.corpus_path
            copy.moniker = self.moniker

        copy._document = self._document.copy(res_opt) if self._document else None

        return copy

    def validate(self) -> bool:
        if not bool(self.corpus_path):
            missing_fields = ['corpus_path']
            logger.error(self.ctx, self._TAG, 'validate', self.at_corpus_path, CdmLogCode.ERR_VALDN_INTEGRITY_CHECK_FAILURE, self.at_corpus_path, ', '.join(map(lambda s: '\'' + s + '\'', missing_fields)))
            return False
        return True

    def visit(self, path_from: str, pre_children: 'VisitCallback', post_children: 'VisitCallback') -> bool:
        # not much to do
        if pre_children and pre_children(self, path_from):
            return False

        if post_children and post_children(self, path_from):
            return True

        return False

File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Cdm/CdmImport.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Cdm
{
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Persistence.CdmFolder.Types;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Microsoft.CommonDataModel.ObjectModel.Utilities.Logging;
    using System;
    using System.Collections.Generic;
    using System.Linq;

    public class CdmImport : CdmObjectSimple
    {
        private static readonly string Tag = nameof(CdmImport);
        /// <summary>
        /// Gets or sets the document that has been resolved for this import.
        /// </summary>
        internal CdmDocumentDefinition Document { get; set; }

        /// <summary>
        /// Gets or sets the import path.
        /// </summary>
        public string CorpusPath { get; set; }

        /// <summary>
        /// Gets or sets the import moniker.
        /// </summary>
        public string Moniker { get; set; }

        /// <summary>
        /// Constructs a CdmImport.
        /// </summary>
        /// <param name="ctx">The context.</param>
        /// <param name="corpusPath">The import path.</param>
        /// <param name="moniker">The import moniker.</param>
        public CdmImport(CdmCorpusContext ctx, string corpusPath, string moniker)
            : base(ctx)
        {
            this.CorpusPath = corpusPath;
            this.Moniker = moniker;
            this.ObjectType = CdmObjectType.Import;
        }

        [Obsolete]
        public override CdmObjectType GetObjectType()
        {
            return CdmObjectType.Import;
        }

        [Obsolete("CopyData is deprecated. Please use the Persistence Layer instead.")]
        public override dynamic CopyData(ResolveOptions resOpt, CopyOptions options)
        {
            return CdmObjectBase.CopyData<CdmImport>(this, resOpt, options);
        }

        /// <inheritdoc />
        public override CdmObject Copy(ResolveOptions resOpt = null, CdmObject host = null)
        {
            if (resOpt == null)
            {
                resOpt = new ResolveOptions(this, this.Ctx.Corpus.DefaultResolutionDirectives);
            }

            CdmImport copy;
            if (host == null)
            {
                copy = new CdmImport(this.Ctx, this.CorpusPath, this.Moniker);
            }
            else
            {
                copy = host as CdmImport;
                copy.Ctx = this.Ctx;
                copy.CorpusPath = this.CorpusPath;
                copy.Moniker = this.Moniker;
            }

            copy.Document = this.Document;
            return copy;
        }

        /// <inheritdoc />
        public override bool Validate()
        {
            if (string.IsNullOrWhiteSpace(this.CorpusPath))
            {
                IEnumerable<string> missingFields = new List<string> { "CorpusPath" };
                Logger.Error(this.Ctx, Tag, nameof(Validate), this.AtCorpusPath, CdmLogCode.ErrValdnIntegrityCheckFailure, this.AtCorpusPath, string.Join(", ", missingFields.Select((s) =>$"'{s}'")));
                return false;
            }
            return true;
        }

        [Obsolete("InstanceFromData is deprecated. Please use the Persistence Layer instead.")]
        public static CdmImport InstanceFromData(CdmCorpusContext ctx, Import import)
        {
            return CdmObjectBase.InstanceFromData<CdmImport, Import>(ctx, import);
        }

        /// <inheritdoc />
        public override bool Visit(string pathFrom, VisitCallback preChildren, VisitCallback postChildren)
        {
            if (preChildren != null && preChildren.Invoke(this, pathFrom))
                return false;
            if (postChildren != null && postChildren.Invoke(this, pathFrom))
                return true;
            return false;
        }
    }
}
